sky/provision/azure/instance.py

```diff
@@ -704,6 +709,30 @@ def _fetch_and_map_status(node, resource_group: str) -> None:
     return statuses
 
 
+def _get_cluster_nsg(network_client: Client, resource_group: str,
+                     cluster_name_on_cloud: str) -> NetworkSecurityGroup:
+    """Retrieve the NSG associated with the given name of the cluster."""
+    list_network_security_groups = _get_azure_sdk_function(
+        client=network_client.network_security_groups, function_name='list')
+    legacy_nsg_name = _LEGACY_NSG_NAME.format(
+        cluster_name_on_cloud=cluster_name_on_cloud)
+    second_legacy_nsg_name = _SECOND_LEGACY_NSG_NAME.format(
+        cluster_name_on_cloud=cluster_name_on_cloud)
+    nsg_name = config_lib.get_nsg_name(
+        resource_group=resource_group,
+        cluster_name_on_cloud=cluster_name_on_cloud)
+    possible_nsg_names = [nsg_name, legacy_nsg_name, second_legacy_nsg_name]
+    for nsg in list_network_security_groups(resource_group):
+        if nsg.name in possible_nsg_names:
+            return nsg
```

Comment:
nit: We should prioritize the latest generation NSG name in the resource group.

Reply:
@Michaelvll I checked if the `list` method we are using supports a way to list the nsg objects following the creation time, but it does not support a way: https://raw.githubusercontent.com/Azure/azure-sdk-for-python/7aeef6f433e98571b0b7624f57d3d89570b94949/sdk/network/azure-mgmt-network/azure/mgmt/network/v2024_01_01/aio/operations/_operations.py

Another approach we can take is to get the list and sort it from our end, but the nsg objects that are returned by `list` method does not contain any attributes related to creation time.
<img width="762" alt="Screenshot 2024-09-10 at 8 58 15 PM" src="https://github.com/user-attachments/assets/e5513c87-19ec-423a-9dd1-ffa00df6b5da">

Reply:
Each cluster / resource group should only have 1 nsg so it's fine to keep it this way